### ISSUE TITLE
Migrate legacy dictionaries without a target_locale set

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#9344]
To automatically migrate legacy dictionaries on AMO to WebExtension format, we need to know their locale - this is because the target locale is now a property in the manifest.

The problem is, some dictionaries don't have the `target_locale` field set, so we have no way of knowing what locale they are for, and therefore can't migrate them to webextension format automatically (the name of the dictionary file in the archive is not enough - it's arbitrary and in some cases does not correspond to the locale).

What shall we do about them ? The options are:
- Have an admin set `target_locale` on those dictionaries. That will make them appear on the language tools page
- Have someone figure out the locale for each of those dictionaries and save it somewhere for the migration script to use
- Ignore them, let the developers of those dictionaries do the migration themselves. The legacy versions would eventually be disabled.

**Comment by diox** [comment]
Also worth asking: now that there is a locale field in the manifest... should we take advantage of that and fill the `target_locale` field automatically when a webextension dictionary is uploaded ? Right now it serves as a de-facto filter, preventing add-ons that have not been reviewed by an admin from being shown in the language tools page (since only admins can change that field). But if that filtering is not deemed necessary we could automatically fill that field from the manifest data.

**Comment by diox** [comment]
cc @jvillalobos to answer both questions

**Comment by jvillalobos** [comment]
> What shall we do about them ?

Do you have a notion of how common it is for the files in the package to be named after the locale? My recollection is that it's standard practice. If we can migrate most of the dictionaries this way, even if it's inaccurate in some cases, I think that's the best way forward. I certainly wouldn't invest anyone's time manually checking all of them. If the files don't buy us much, ignoring those add-ons for the migration is acceptable.

> should we take advantage of that and fill the target_locale field automatically when a webextension dictionary is uploaded ?

I suppose that's okay. It's not something we have to deal with often and it'd be good to automate it. We could run into a problem if many dictionaries are submitted for the same locale, but that's unlikely to happen, and not a bad problem to have.

**Comment by diox** [comment]
Out of the 136 dictionaries with `target_locale` set at the moment, 79 have a filename that does not directly match a known Firefox locale (as pulled from [the Product-Details API](https://wiki.mozilla.org/Release_Management/Product_details#languages.json)). Most of them look like a locale, just not one that we know about.

Expanding to all dictionaries, including those without `target_locale`, out of 245 dictionaries, 166 have a filename that does not directly match a known Firefox locale.

**TL;DR:** the filename of the `.dic` file inside the archive is a poor indicator of the actual locale.

**Comment by flodolo** [comment]
Can you publish somewhere a list of these filenames? I wonder how much of that is due to differences like `it` (locale code shipping in Firefox) vs `it_IT`.

**Comment by diox** [comment]
The full list: 
```python
['es-CO',
 'sc-GSC',
 'cu',
 'ta-TA',
 'ak-GH',
 'venetan',
 'et-US',
 'tr-TR',
 'ga-Latg-IE',
 'gn_PY',
 'or-IN',
 'fr-inclusif',
 'fr-classic',
 'tt',
 'mg_MG',
 'st-ZA',
 'be-latn',
 'el-GR',
 'pt-PT-preAO',
 'yi',
 'ltg_LV',
 'Papiamento',
 'it_IT',
 u'bb-BK',
 'rw-RW',
 'Moore-Extended',
 'as-IN',
 'en-x-simple',
 'ru_RU',
 'Uyghur',
 'kv',
 'mi-x-Tai Tokerau',
 'fa-IR',
 'neolatino',
 'nr-ZA',
 'nn',
 'ka-GE',
 'mr-IN',
 'ia-ia',
 'roa-ES-val',
 'bal',
 'mn-MN',
 'de-BG',
 'el_GR',
 'gl_ES',
 'pt-BR-antigo',
 'ie',
 'hyph_pt-BR',
 'my_MM',
 'ss-ZA',
 'Persian (Nowdabira)',
 'ru_RU',
 'lv_LV',
 'ts-ZA',
 'cy_GB',
 'szl',
 'de-DE-1901',
 'es_VE',
 'ku-CKB',
 'ta_TA',
 'csk-SN',
 'en-Medical',
 'sr-RS-Cyrl',
 'tn-ZA',
 'el-GR',
 'bg_IV',
 'hr-HR',
 'sa_IN',
 'mt',
 'MedSpell',
 'Uyghur (Latin)',
 'Uyghur (Cyrillic)',
 'ro_RO-ante1993',
 'te_IN',
 'fo_FO',
 've-ZA',
 'BIH-CRO-SRB',
 'mh-MH',
 'tk_TM',
 'km_KH',
 u'\u041c\u0430\u0440\u0438\u0439',
 'ur-IN',
 'OpenMedSpel',
 'gl_ES-volga',
 'hsb-DE',
 'haw-US',
 'ny_MW',
 'ku-TR',
 'mk-MK-Cyrl',
 'ko-KR',
 u'\u0423\u0434\u043c\u0443\u0440\u0442',
 'nso-ZA',
 'bm-ML',
 'af-ZA',
 'fi-FI',
 'nb',
 'my_MM',
 'vi-DauCu',
 'es_NI',
 'dyo-SN',
 'so_SO',
 'zu-ZA',
 'ru_RU',
 'ms-MY',
 'si-LK',
 'tpi-PG',
 'gd-GB',
 'en-bg',
 'Dhivehi',
 'et-EE',
 'kz',
 'el_GR-en_US',
 'Chechen-Noxchi',
 'English & Greek',
 'oc-FR',
 'Srpski (Latinica)',
 'en-x-basic',
 'be-classic',
 'ht-HT',
 'en-ru',
 'Persian_Alefba_ye_2om',
 'cy_GB',
 'be-tarask',
 'eo-EO',
 'de-DE-1901',
 'ca-ES-valencia',
 'ia-ia',
 'en',
 'en-bg',
 'an_ES',
 'ru-English',
 'sv',
 'nds-DE',
 'en-x-simpler',
 'km_KH',
 'lb-LU',
 'nb',
 'hus-MX',
 'ru-petr1708',
 'ru-Deutsch',
 'cv',
 'ml_IN',
 'Papiamentu',
 'am_ET',
 'mn-MN',
 'az-Latn-AZ',
 'ln-CD',
 'eu_ES',
 'sw_TZ',
 'mi-NZ',
 'hil',
 'xh-ZA',
 'fr-FR-2014',
 u'uk-UA',
 'Kaszebsczi',
 'om_ET',
 'ru-uz',
 'son-ML',
 'multilanguages',
 'lo_LA',
 'ny-MW',
 'Cestina',
 'nko-GN',
 'ha-GH',
 'uz-UZ',
 'bo']
```

**Comment by diox** [comment]
(Note that, when trying to match to known locales, I also tried to massage the name, transforming underscores into dashes, that kind of thing. It didn't really help much)

**Comment by diox** [comment]
It would help, yeah. That still leaves 69 that don't match:

```python
['sc-GSC',
 'cu',
 'venetan',
 'tt',
 'pt-PT-preAO',
 'yi',
 'Papiamento',
 u'bb-BK',
 'Moore-Extended',
 'en-x-simple',
 'Uyghur',
 'kv',
 'neolatino',
 'nn',
 'roa-ES-val',
 'bal',
 'pt-BR-antigo',
 'ie',
 'hyph_pt-BR',
 'Persian (Nowdabira)',
 'szl',
 'csk-SN',
 'en-Medical',
 'mt',
 'MedSpell',
 'Uyghur (Latin)',
 'Uyghur (Cyrillic)',
 'fo_FO',
 'BIH-CRO-SRB',
 'mh-MH',
 'tk_TM',
 u'\u041c\u0430\u0440\u0438\u0439',
 'OpenMedSpel',
 'haw-US',
 'ny_MW',
 u'\u0423\u0434\u043c\u0443\u0440\u0442',
 'nb',
 'dyo-SN',
 'so_SO',
 'tpi-PG',
 'en-bg',
 'Dhivehi',
 'kz',
 'Chechen-Noxchi',
 'English & Greek',
 'Srpski (Latinica)',
 'en-x-basic',
 'ht-HT',
 'en-ru',
 'Persian_Alefba_ye_2om',
 'en',
 'en-bg',
 'sv',
 'nds-DE',
 'en-x-simpler',
 'lb-LU',
 'nb',
 'hus-MX',
 'cv',
 'Papiamentu',
 'am_ET',
 'hil',
 'Kaszebsczi',
 'om_ET',
 'multilanguages',
 'ny-MW',
 'Cestina',
 'nko-GN',
 'bo']
```

**Comment by flodolo** [comment]
I think it would be OK to exclude these from a migration.

Some belong to actual locales that we don't ship (e.g. lb), others either have made up names, or look wrong. In case, would it be possible to reach out to the author and explain what they need to do to migrate?

**Comment by jvillalobos** [comment]
Agree on 69 unmatched being okay. And yes, we can try reaching out to those devs about migrating on their own.

**Comment by AlexandraMoga** [comment]
@diox is the following test approach correct:
* submit several legacy dictionaries
* make sure that they do not have a `target_locale` in Admiin

after the migration script is run

* check the `manifest.json` to make sure that a `target_locale` is present
* make sure that the `target_locale` field was automatically populated in Admin

Let me know if something is missing or is inaccurate.

**Comment by diox** [comment]
@AlexandraMoga that's correct. In addition, since the `target_locale` will be determined from the name of the `.dic` file, you can try submitting legacy dicts with the following `.dic` files inside:
- One with a language we support directly, for instance `fr.dic`
- One with a language variant we don't support, for instance `fr-something.dic`. It should result in a `fr` locale, ignoring the `-something` part.
- One with a language we definitely can't recognize. It should be skipped.

You can also play with `_` instead of `-` in the filename, as well as lower/upper case combinations, we should handle that. Let me know when you've uploaded your test dicts and I'll run the migration again.

**Comment by diox** [comment]
Re-opening because of this issue:
```
Uncaught exception:   
File "/usr/local/lib/python2.7/site-packages/celery/app/trace.py", line 375, in trace_task
     R = retval = fun(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/celery/app/trace.py", line 632, in __protected_call__
     return self.run(*args, **kwargs)
File "/data/olympia/src/olympia/amo/decorators.py", line 109, in wrapper
     return f(*args, **kw)   
File "/data/olympia/src/olympia/addons/tasks.py", line 708, in migrate_legacy_dictionaries_to_webextension
     log('Migrating dictionary %d raised %s', addon.pk, exc.message)
<type 'exceptions.TypeError'> TypeError("'AmoLoggerAdapter' object is not callable",)
```

**Comment by AlexandraMoga** [comment]
Verified on AMO -dev with FF62, WIn10x64

Here are the migration results:

**1. Migrated dictionaries (`target_locale` recognized)**
- https://addons-dev.allizom.org/en-US/firefox/addon/norsk-bokm%C3%A5l-ordliste-nb-no/ - `nb-no` is a supported locale
- https://addons-dev.allizom.org/en-US/firefox/addon/german-dictionary-de-schweiz/ - `de` was identified in the name of the `.dic` file as a supported locale
- https://addons-dev.allizom.org/en-US/firefox/addon/dizionario-italiano-it_it/ - `it_IT` was interpreted as `it` which is a supported locale

**2. Not migrated (`target_locale` not supported as per https://github.com/mozilla/addons-server/issues/9344#issuecomment-418711868 )**
- https://addons-dev.allizom.org/en-US/firefox/addon/new-zealand-english-dictionary/ - `EN-NZ` not identified as a supported locale
- https://addons-dev.allizom.org/en-US/firefox/addon/norsk-bokm%C3%A5l-ordliste-nb/ - `nb` is listed as a unsupported locale
- https://addons-dev.allizom.org/en-US/firefox/addon/svenska-fria-ordlistan-sv-fi/ - `sv` or `sv-FI` not listed as supported locales
- https://addons-dev.allizom.org/en-US/firefox/addon/diccionari-ortogr%C3%A0fic-valenci%C3%A0/ -`roa-ES-val`listed as unsupported locale

**3. Needs investigation (not migrated, `target_locale` not listed as unsupported)**
- https://addons-dev.allizom.org/en-US/firefox/addon/hrvatski-rje%C4%8Dnik-hr/ - the name of the `.dic` file was `hr`; `hr-HR` is listed as a supported locale, so I've expected this dictionary to be migrated
- https://addons-dev.allizom.org/en-US/firefox/addon/northern-sotho-/ - the name of the `.dic` file was `nso_ZA`; `nso-ZA` is listed as a supported locale,

@diox for the dictionaries that were successfully migrated, I've looked in Admin tools but the Target Locale field was not automatically populated.
Please have a look also at the dictionaries mentioned at point 3.

**Comment by diox** [comment]
> https://addons-dev.allizom.org/en-US/firefox/addon/hrvatski-rje%C4%8Dnik-hr/ - the name of the .dic file was hr; hr-HR is listed as a supported locale, so I've expected this dictionary to be migrated

To be expected, we can't match `hr` to `hr-HR` automatically because there is no link between the 2 in our settings.

> https://addons-dev.allizom.org/en-US/firefox/addon/northern-sotho-/ - the name of the .dic file was nso_ZA; nso-ZA is listed as a supported locale,

Where is it listed? It's not in addons-server settings.

> @diox for the dictionaries that were successfully migrated, I've looked in Admin tools but the Target Locale field was not automatically populated.

I'll have a look at this.

**Comment by flodolo** [comment]
We also don't ship nso anywhere (we might have tried at Firefox OS time).

**Comment by AlexandraMoga** [comment]
> Where is it listed? It's not in addons-server settings.

I've seen `nso-ZA` listed in your first locales list from https://github.com/mozilla/addons-server/issues/9344#issuecomment-418706956, but then it wasn't included among the 69 unsupported locales left. Maybe I've read your list the wrong way.

**Comment by diox** [comment]
Not sure why it's not in the remaining unsupported locales, maybe I've messed something up when generating it.

**Comment by AlexandraMoga** [comment]
Since we have #9522 filed as a follow up to handle the automatic population of target_locales, I'm going to mark this issue as verified fixed on AMO -dev with FF62, Win10x64.
